image_classification/resnet50v2_nchw.js

```diff
@@ -0,0 +1,165 @@
+    this.inputOptions = {
+      mean: [0.485, 0.456, 0.406],
+      std: [0.229, 0.224, 0.225],
+      norm: true,
+      inputLayout: 'nchw',
+      labelUrl: './labels/labels1000.txt',
+      inputDimensions: [1, 3, 224, 224],
+    };
+  }
+
+  async buildConv_(input, name, stageName, options = undefined) {
+    let prefix = '';
+    if (stageName !== '') {
+      prefix = this.weightsUrl_ + 'resnetv24_stage' + stageName + '_conv' +
+          name;
+    } else {
+      prefix = this.weightsUrl_ + 'resnetv24_conv' + name;
+    }
+    const weightName = prefix + '_weight.npy';
+    const weight = await buildConstantByNpy(this.builder_, weightName);
+    return this.builder_.conv2d(input, weight, options);
+  }
+
+  async buildBatchNorm_(input, name, stageName, relu = true) {
+    let prefix = '';
+    if (stageName !== '') {
+      prefix = this.weightsUrl_ + 'resnetv24_stage' + stageName +
+          '_batchnorm' + name;
+    } else {
+      prefix = this.weightsUrl_ + 'resnetv24_batchnorm' + name;
+    }
+    const scaleName = prefix + '_gamma.npy';
+    const biasName = prefix + '_beta.npy';
+    const meanName = prefix + '_running_mean.npy';
+    const varName = prefix + '_running_var.npy';
+    const scale = await buildConstantByNpy(this.builder_, scaleName);
+    const bias = await buildConstantByNpy(this.builder_, biasName);
+    const mean = await buildConstantByNpy(this.builder_, meanName);
+    const variance = await buildConstantByNpy(this.builder_, varName);
+    const batchNorm = this.builder_.batchNormalization(
+        input, mean, variance, {scale: scale, bias: bias});
+    if (relu) {
+      return this.builder_.relu(batchNorm);
+    }
+    return batchNorm;
+  }
+
+  async buildGemm_(input, name) {
+    const prefix = this.weightsUrl_ + 'resnetv24_dense' + name;
+    const weightName = prefix + '_weight.npy';
+    const weight = await buildConstantByNpy(this.builder_, weightName);
+    const biasName = prefix + '_bias.npy';
+    const bias = await buildConstantByNpy(this.builder_, biasName);
+    const options = {c: this.builder_.reshape(bias, [1, -1]), bTranspose: true};
+    return this.builder_.gemm(input, weight, options);
+  }
+
+  async buildLinearBottleneck1_(input, stageName, stride = 1) {
+    const strides = [stride, stride];
+
+    const batchNorm0 = await this.buildBatchNorm_(input, '0', stageName);
+    const conv0 = await this.buildConv_(batchNorm0, '0', stageName);
+    const batchNorm1 = await this.buildBatchNorm_(conv0, '1', stageName);
+    const conv3x3 = await this.buildConv_(
+        batchNorm1, '1', stageName, {padding: [1, 1, 1, 1], strides});
+    const batchNorm2 = await this.buildBatchNorm_(conv3x3, '2', stageName);
+    const conv1 = await this.buildConv_(batchNorm2, '2', stageName);
+    const conv2 = await this.buildConv_(batchNorm0, '3', stageName, {strides});
+
+    return this.builder_.add(conv1, conv2);
+  }
+
+  async buildLinearBottleneck2_(input, stageName, nameIndex) {
```

Comment:
According to [paper](https://arxiv.org/pdf/1512.03385.pdf) and [training script](https://github.com/onnx/models/blob/master/vision/classification/resnet/train_resnet.ipynb), I'd suggest to combine `buildLinearBottleneck1_` and `buildLinearBottleneck2_` into `buildBottleneckV2_` function such as
```js
async buildBottleneckV2_(input, stageName, nameIndex, downsample = false, stride = 1) {
    let residual = input;
    const strides = [stride, stride];
    const bn1 = await this.buildBatchNorm_(input, nameIndex, stageName);
    if (downsample) {
      residual = await this.buildConv_(bn1, nameIndex + 3, stageName, {strides});
    } else {
      nameIndex = nameIndex + 1;
    }
    const conv1 = await this.buildConv_(bn1, nameIndex, stageName);
    const bn2 = await this.buildBatchNorm_(conv1, nameIndex + 1, stageName);
    const conv2 = await this.buildConv_(bn2, nameIndex + 2, stageName, {padding: [1, 1, 1, 1], strides});
    const bn3 = await this.buildBatchNorm_(conv2, nameIndex + 2, stageName);
    const conv3 = await this.buildConv_(bn3, nameIndex + 3, stageName);

    return this.builder_.add(conv3, residual);
  }
```

There might be syntax errors in my code, but you got the idea. Thanks!

Reply:
```js
async buildBottlenectV2_(input, stageName, nameIndex, downsample = false, stride = 1) {
    let residual = input;
    const strides = [stride, stride];

    const batchNorm0 = await this.buildBatchNorm_(input, nameIndex, stageName);
    let conv0;
    if (downsample) {
      conv0 = await this.buildConv_(batchNorm0, nameIndex, stageName);
    } else {
      conv0 = await this.buildConv_(batchNorm0, nameIndex + 1, stageName);
    }
    const batchNorm1 = await this.buildBatchNorm_(conv0, nameIndex + 1, stageName);
    let conv3x3;
    if (downsample) {
      conv3x3 = await this.buildConv_(
        batchNorm1, nameIndex + 1, stageName, {padding: [1, 1, 1, 1], strides});
    } else {
      conv3x3 = await this.buildConv_(
        batchNorm1, nameIndex + 2, stageName, {padding: [1, 1, 1, 1]});
    }
    const batchNorm2 = await this.buildBatchNorm_(conv3x3, nameIndex + 2, stageName);
    if (downsample) {
      const conv1 = await this.buildConv_(batchNorm2, nameIndex + 2, stageName);
      const conv2 = await this.buildConv_(batchNorm0, nameIndex + 3, stageName, {strides});
      return this.builder_.add(conv1, conv2);
    } else {
      const conv1 = await this.buildConv_(batchNorm2, nameIndex + 3, stageName);
      return this.builder_.add(conv1, input);
    }
  }
```
@huningxin, I created a combined function to view how many different between these two `bottleneck`, it looks a bit complicated, especially for those un-aligned `nameIndex`.

Reply:
@Honry , I think my sample code is equivalent to yours for:



> ```js
>  if (downsample) {
>       conv0 = await this.buildConv_(batchNorm0, nameIndex, stageName);
>     } else {
>       conv0 = await this.buildConv_(batchNorm0, nameIndex + 1, stageName);
>     }
> ```

You can reset the base of `nameIndex` when `downsample` is false, e.g. `nameIndex = nameIndex + 1;`



> ```js
> if (downsample) {
>       conv3x3 = await this.buildConv_(
>         batchNorm1, nameIndex + 1, stageName, {padding: [1, 1, 1, 1], strides});
>     } else {
>       conv3x3 = await this.buildConv_(
>         batchNorm1, nameIndex + 2, stageName, {padding: [1, 1, 1, 1]});
>     }
> ```

You can make sure if `downsample` is false, the `stride` is always 1.



> ```js
> if (downsample) {
>       const conv1 = await this.buildConv_(batchNorm2, nameIndex + 2, stageName);
>       const conv2 = await this.buildConv_(batchNorm0, nameIndex + 3, stageName, {strides});
>       return this.builder_.add(conv1, conv2);
>     } else {
>       const conv1 = await this.buildConv_(batchNorm2, nameIndex + 3, stageName);
>       return this.builder_.add(conv1, input);
>     }
> ```

You can set the `residual = await this.buildConv_(bn1, nameIndex + 3, stageName, {strides});` when `downsample` is true, otherwise set the `residual = input`, then add it into the conv3 at the last as `return this.builder_.add(conv3, residual);`.

Reply:
The problem is the `nameIndex` in common `const batchNorm1 = await this.buildBatchNorm_(conv0, nameIndex + 1, stageName);` and `const batchNorm2 = await this.buildBatchNorm_(conv3x3, nameIndex + 2, stageName);` declarations  could not be align when we set `nameIndex = nameIndex + 1;` in `downsample` is false.

Reply:
Thus we need more `if...else...` for `batchNorm1` and `batchNorm2`.

Reply:
The architecture of residual unit is basically same with different shortcut connect (downsample or not). So it would make sense to use one code flow to build it. If the problem is only about the nameIndex, probably we can supply them, e.g.

```js
async buildBottleneckV2_(input, stageName, nameIndices, downsample = false, stride = 1) {
  let residual = input;
  const strides = [stride, stride];
  const bn1 = await this.buildBatchNorm_(input, nameIndices[0], stageName);
  const conv1 = await this.buildConv_(bn1, nameIndices[1], stageName);
  const bn2 = await this.buildBatchNorm_(conv1, nameIndices[2], stageName);
  const conv2 = await this.buildConv_(bn2, nameIndices[3], stageName, {padding: [1, 1, 1, 1], strides});
  const bn3 = await this.buildBatchNorm_(conv2, nameIndices[4], stageName);
  const conv3 = await this.buildConv_(bn3, nameIndices[5], stageName);
  if (downsample) {
    residual = await this.buildConv_(bn1, nameIndices[6], stageName, {strides});
  }
  return this.builder_.add(conv3, residual);
}

```